### PR TITLE
wip

### PR DESCRIPTION
**Pull request by arnetheduck**
this pr is a playground for orienting the protocol towards pull, with a bunch of simplifications thrown in for free:

* use `StreamTransport` instead of `AsyncStream` to get rid of a seemingly redundant abstraction / buffer
* use future to transport data between layers

it still needs the following simplifications:
* `SimpleStream` currently has a buffer - this is needed only because the interface supports multiple `readUntil`, `readLine` etc methods - all of them are redundant and should be removed in favour of a single `readHandler` at each level - the rest can be written as helpers outside - then, the buffer in simplestream could be removed and all that's needed would be a buffer in whatever is doing the reading - for example secio needs a buffer to hold a full frame - this can easily be implemented using the new `readMessage` api in chronos

The following things are broken:
* exception handling (it was broken and inconsistent before, and continues to be so, albeit in different ways)
* there's a synchronization issue in `mplex`: when someone wants to read data from an mplex channel, they plant a future in `readerFuture` which will be completed by the mplex pushTo - if there's no future, the data will be dropped - this means that the planting of the future must happen before the mplex loop reads the data and tries to push it - this leads to a race condition between whoever registers the future and the mplex reader loop, because if the consumer of data does any async operation at all, the mplex loop might happen "first"

### CODE REVIEW

File: tests/testmplex.nim [hunk 1]
```diff
@@ -402,7 +402,7 @@ suite "Mplex":
       proc writeHandler(data: seq[byte]) {.async, gcsafe.} = discard
       let chann = newChannel(1, newConnection(newBufferStream(writeHandler)), true)
       await chann.closedByRemote()
-      await chann.pushTo(@[byte(1)])
+      chann.pushTo(@[byte(1)])
```
Comment (cheatfate) [inline tests/testmplex.nim]: I think `await` here is needed.

**Comment by arnetheduck** [comment]
Expanding on the design issues:

* When data arrives, the mplex read loop plucks it off the socket and must decide what to do with the data. If it receives a message to create a new channel, that new channel needs a consumer that will handle the data and decide if it's worthwhile to keep or not
* If there is nobody interested, the best thing mplex can do is either:
  * store the data - this has buffer or memory overflow problems if nobody is periodically clearing the buffer
  * remove the channel - if any data is dropped, the channel becomes useless, so we might as well drop it
* if there is somebody interested for each channel, there is no problem: data is delivered directly to the interested party that takes care of it. however, this PR is flawed in that it must re-insert its interest for data before handing control back to chronos, or a race condition happens: the next iteration of the mplex read loop may run before the registration has happened - basically, the delivery of the data needs to be synchronous.
  * this synchronous requirement can be relaxed by introducing a buffer, similar to what there is now - obviously that leads to some of the same issues we have now with potential for buffer overflow - but a key problem with that is that there exists no buffer size that's guaranteed to be good for all protocols - they all vary in how much buffer space they would need

the situation is similar to data arriving over TCP and the ways the OS handles this:
* the model in this PR is a polling `recv` model - the os places data from the socket in a buffer and waits for the reader to pick it up. when the data is picked up, the OS tells the TCP sender (with an ack) that it's ready to handle more data - mplex seems to lack this capability - there's no way to stop the flood except not listening on the socket any more
* the currently implemented push model is not too different from epoll with edge-triggered events - however, there's still no way to tell `mplex` to tell the sending side to "pause" a particular stream - this is what the OS uses to let the other sockets communicate while one is blocked / slow

all in all, the lack of pausing a single stream in `mplex` seems fundamental - when unknown data is arriving, either one has to:
* pause all channels
* drop the single channel
* handle the data "synchronously" or pseudo-synchronously with a more intelligent predicate than max buffer size

the last option is in a way the most appealing one: how much data to to buffer really depends on the final consumer of the data and how its framing works - if it's reading messages, the right amount to buffer before deciding that the channel is bad is one message - other formats will need other ways to check the validity of the data being received, meaning we need a way to specify a predicate that decides what to do - the new `readMessage` API in chronos does this with a predicate callback, but is still not sufficient in that it's still vulnerable to the async race condition.

the other way is to not use mplex and find some other multiplexer that acks to the sender how much data a single channel has processed (basically, TCP inside the channel).

**Comment by dryajov** [comment]
> Expanding on the design issues:
> 
> * When data arrives, the mplex read loop plucks it off the socket and must decide what to do with the data. If it receives a message to create a new channel, that new channel needs a consumer that will handle the data and decide if it's worthwhile to keep or not
> * If there is nobody interested, the best thing mplex can do is either:
>   
>   * store the data - this has buffer or memory overflow problems if nobody is periodically clearing the buffer
>   * remove the channel - if any data is dropped, the channel becomes useless, so we might as well drop it

All of this is implicitely guaranteed by backpressure, which is the natural mechanism to deal with this sort of issues, the buffer should hold at most the amount of the last few bytes from the last read. In our current implementation, the internal buffer is intentionally left at `1024` bytes, to ensure that we only hold a very small amount of data across all the channels, in this implementation it should be at most the the max amount of the mplex channel (1mb) which is the implicit buffer passed to `pushTo`. 

The reason you don't see that behavior in this implementation is because it essentially has no backpressure mechanism, because `pushTo` will non-discretionary dump data as it comes in (and potentially lock up the loop all together because nothing else will be awoken until all writes have stopped, so this is a potential vulnerability as well). In order to add backpressure `pushTo` needs to `await` until there is a consumer, so it has to be async. This is fine because `mplex` is a *serial* mechanism and by design only one channel writes/reads to the underlying socket at a time, and it does so in at most max message size chunks, so if there is one consumer that's slow the whole pipe should slow down and that's backpressure.

> * if there is somebody interested for each channel, there is no problem: data is delivered directly to the interested party that takes care of it. however, this PR is flawed in that it must re-insert its interest for data before handing control back to chronos, or a race condition happens: the next iteration of the mplex read loop may run before the registration has happened - basically, the delivery of the data needs to be synchronous.
>   
>   * this synchronous requirement can be relaxed by introducing a buffer, similar to what there is now - obviously that leads to some of the same issues we have now with potential for buffer overflow - but a key problem with that is that there exists no buffer size that's guaranteed to be good for all protocols - they all vary in how much buffer space they would need

We might need a buffer but for entirely different reasons. We might need it, if we want to hold less bytes than the max size of the incoming message, and that was the reason it's currently present in `BufferStream`. Honestly, I don't think we need that anymore, the implicit buffer should be enough and the 1mb max channel size is reasonable, tho it does decrease our tolerance to channel spamming, but not by a large margin.

> 
> the situation is similar to data arriving over TCP and the ways the OS handles this:
> 
> * the model in this PR is a polling `recv` model - the os places data from the socket in a buffer and waits for the reader to pick it up. when the data is picked up, the OS tells the TCP sender (with an ack) that it's ready to handle more data - mplex seems to lack this capability - there's no way to stop the flood except not listening on the socket any more
> * the currently implemented push model is not too different from epoll with edge-triggered events - however, there's still no way to tell `mplex` to tell the sending side to "pause" a particular stream - this is what the OS uses to let the other sockets communicate while one is blocked / slow

Again, this is mentioned in my previous paragraph but it's worth emphasizing that the mechanism is there and it is backpressure, the problem with this PR is that it lost backpressure the moment `pushTo` became non-async.

> 
> all in all, the lack of pausing a single stream in `mplex` seems fundamental - when unknown data is arriving, either one has to:
> 
> * pause all channels
> * drop the single channel
> * handle the data "synchronously" or pseudo-synchronously with a more intelligent predicate than max buffer size
> 
> the last option is in a way the most appealing one: how much data to to buffer really depends on the final consumer of the data and how its framing works - if it's reading messages, the right amount to buffer before deciding that the channel is bad is one message - other formats will need other ways to check the validity of the data being received, meaning we need a way to specify a predicate that decides what to do - the new `readMessage` API in chronos does this with a predicate callback, but is still not sufficient in that it's still vulnerable to the async race condition.
> 
> the other way is to not use mplex and find some other multiplexer that acks to the sender how much data a single channel has processed (basically, TCP inside the channel).

To summarize, this PR is a great start and it gets a lot of things right, thanks @arnetheduck! 

The one thing it's missing is a backpressure which was lost when `pushTo` became synchronous, we need to make it `async` again and add a "nicer" mechanism to synchronize it, such as `AsyncEnvent` with a data payload, that would make `pushTo` await until a consumer read the data, however chronos currently lacks the ability to pass data with `AsyncEvent` and @cheatfate is against making that modification, so an alternative is to use the `AsyncQueue` with one element. This queue/event will replace the current `Future` and simply read the last available message as soon it becomes ready, not more `nil` checks and sync issues. 

I'll push my changes shortly and we can keep iterating on it.

**Comment by dryajov** [comment]
Commit fe8c2ad has the modifications to bring backpresure back, most tests are passing except for a few that look unrelated to this change.

**Comment by arnetheduck** [comment]
> All of this is implicitely guaranteed by backpressure, which is the natural mechanism to deal with this sort of issues

the consequence of mplex lacking per-channel signalling mechanisms / ack is that the backpressure de facto gets applied to _all_ channels, not just the one that is faulty / slow / whatever. In consequence, the transport stops being read completely - the current design however looks like (and introduces the complexity of) having split backpressure accounting for each channel - this is an impedance mismatch, to use a familiar expression :) 

another way to put the above is that with mplex, the backpressure handling that the socket itself already offers is superior in simplicity, memory usage and performance.

**Comment by dryajov** [comment]
> > All of this is implicitely guaranteed by backpressure, which is the natural mechanism to deal with this sort of issues
> 
> the consequence of mplex lacking per-channel signalling mechanisms / ack is that the backpressure de facto gets applied to _all_ channels, not just the one that is faulty / slow / whatever. In consequence, the transport stops being read completely - the current design however looks like (and introduces the complexity of) having split backpressure accounting for each channel - this is an impedance mismatch, to use a familiar expression :)
> 
> another way to put the above is that with mplex, the backpressure handling that the socket itself already offers is superior in simplicity, memory usage and performance and there exists a design (that still remains to be created) that matches this.

I agree, but this is by design, mplex is a very simple muxing mechanism that has this tradeoffs and all mplex implementations share this limitation, this is in part why it's so simple. There are alternatives to it like `yamux` which I believe have a more sophisticated flow control mechanism.

**Comment by dryajov** [comment]
> the consequence of mplex lacking per-channel signalling mechanisms / ack is that the backpressure de facto gets applied to all channels, not just the one that is faulty / slow / whatever. In consequence, the transport stops being read completely - the current design however looks like (and introduces the complexity of) having split backpressure accounting for each channel - this is an impedance mismatch, to use a familiar expression :)

A counterpoint to this is that by having backpressure per channel you now have to deal with all this corner cases you described, but more importantly you quickly run into the dilemma of what to do with the data that has already arrived? Do you drop it? Why? If you drop it, is there a mechanism to signal the other side that it has been dropped so that we don't violate TCP semantics? This quickly gets complex and hence `mplex` and it's simple model.

**Comment by arnetheduck** [comment]
> A counterpoint to this is that by having backpressure per channel you now have to deal with all this corner cases you described, 

I'm not saying we should drop mplex and use yamux - I'm saying that as long as we use mplex, our mplex implementation should not have code that looks like per-channel backpressure because that's not a feature that mplex offers.

> but more importantly you quickly run into the dilemma of what to do with the data that has already arrived? 

you keep a window and you stop sending acks for it. the other side stops sending data - just like TCP.

**Comment by arnetheduck** [comment]
> The current design blocks all channels

yes, this is what I meant by impedance mismatch - it looks like it's per-channel when in reality it's not - this makes it harder to audit and reason about

**Comment by dryajov** [comment]
> yes, this is what I meant by impedance mismatch - it looks like it's per-channel when in reality it's not - this makes it harder to audit and reason about

OK, yes I can see that. 

But really we should implement [`yamux`](https://github.com/hashicorp/yamux/blob/master/spec.md) once we have this one working and have that be our default `multiselect`/`multistream-select` is here.

**Comment by arnetheduck** [comment]
> once we have this one working

I'd wait for quic / multistream 2 (or was it select) and make the call then. this is good enough for now, if we solve the issues with a clean implementation and some sane limits on channel opening etc.

**Comment by dryajov** [comment]
Well, `multiselect` would help to negotiate the best multiplexer for sure, but quiq really won't make much difference from mixer stand point and `yamux` is another of those standards that's both in eth2 and libp2p, so unless that changes we'll _have_(?) to implement it, tho we can just get away with just mplex.

**Comment by arnetheduck** [comment]
`QUIC` has multi-stream capabilities built in, with each stream being individually flow-controlled, ie at that point it's doubtful you need yamux or mplex - there  was interest to use it for ETH2 but we decided it was too far off to introduce in this version of the spec.

**Comment by cheatfate** [comment]
I'm not sure, but i think for QUIC libp2p uses native QUIC's multiplex feature.

**Comment by dryajov** [comment]
> `QUIC` has multi-stream capabilities built in, with each stream being individually flow-controlled, ie at that point it's doubtful you need yamux or mplex - there was interest to use it for ETH2 but we decided it was too far off to introduce in this version of the spec.

Yeah, you right quic does have multiplexing built in, yamux can of course wait.

**Comment by sinkingsugar** [comment]
I really like this PR changes, it kinda is what I was hoping to find when implementing noise.
In fact I would like to wrap it up once #90 is merged ( locally I have it in a merged branch but easy to get out of sync ).

File: tests/testtransport.nim [hunk 2]
```diff
@@ -119,15 +119,20 @@ suite "TCP transport":
     proc testListenerDialer(): Future[bool] {.async.} =
       let ma: MultiAddress = Multiaddress.init("/ip4/0.0.0.0/tcp/0")
       proc connHandler(conn: Connection): Future[void] {.async, gcsafe.} =
-        let msg = await conn.read(6)
+        # TODO: chronos ``read`` is broken!
```
Comment (sinkingsugar) [inline tests/testtransport.nim]: I had issues with this too, is it really broken tho ? or just mis-used?

**Comment by sinkingsugar** [comment]
I localized most of the issues with the tests, they come from here:
https://github.com/status-im/nim-libp2p/blob/b1a34f478efe6d1212cfccecaf4dd8bf89fc2b1f/libp2p/muxers/mplex/mplex.nim#L139-L143

This call back is generally called in the context of the next test...

This fixes the issues:
```nim
method close*(m: Mplex) {.async, gcsafe.} =
  trace "closing mplex muxer"
  try:
    await allFutures(@[allFutures(toSeq(m.remote.values).mapIt(it.reset())),
                      allFutures(toSeq(m.local.values).mapIt(it.reset()))])
  except LPStreamIncorrectError:
    trace "transport was already closed"
```

Not sure tho in general if I can call this a fix or just a hack

**Comment by arnetheduck** [comment]
#163 contains the last few useful bits from this branch
